#26346: docker volume inspect doesn't do partial ID matching

bbayani opened the pull request:
fixes #26346

**\- What I did**
Implemented code for performing partial ID matching for docker volumes (ID is name for volumes)

**\- How I did it**
Added method GetVolumesByName. It iterates over all volumes in daemon volume store and returns the list of volumes whose names match with passed prefix

**\- How to verify it**
1. Build a developer container with latest docker/master and this PR
2. Create binaries in container with hack/make.sh binary
3. Copy binaries in /usr/bin
4. Start docker daemon
5. Create docker volume with "docker volume create test1"
6. Try docker volume inspect with complete name
docker volume inspect test1
docker inspect test1
7. Try docker volume inspect with first part of volume name
docker volume inspect te
docker inspect te
8. Try volume inspect with non existing volume name and ensure you get error
docker volume inspect abc -> no such volume: abc
docker inspect abc -> no such container/network/volume/service 
9. If more and one volumes with same first initial characters exists, inspect will be performed for first one.

**\- Description for the changelog**
1. Removed a VolumeInspect method from daemon/inspect.go
2. Added this method in daemon/volumes.go
3. Added new method getVolumesByName in daemon/volumes.go which does partial name matching and returns list of matching volumes
4. Replaced method call in VolumeInspect that looks-up volume with complete name, with new method getVolumesByName.  

**\- A picture of a cute animal (not mandatory but encouraged)**
![adorable-giraffe-family](https://cloud.githubusercontent.com/assets/10231863/19310259/dfb593d0-90a6-11e6-9c01-d7a49064461f.jpg)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "volume_id_matching" git@github.com:bbayani/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
Thanks @bbayani - looks like there's some linting issues;

```
06:19:16 ---> Making bundle: validate-lint (in bundles/1.13.0-dev/validate-lint)
06:19:17 Errors from golint:
06:19:17 daemon/volumes.go:29:6: exported type VolumeWalker should have comment or be unexported
06:19:17 daemon/volumes.go:352:1: comment on exported method Daemon.WalkVolumes should be of the form "WalkVolumes ..."
06:19:17 
06:19:17 Please fix the above errors. You can test via "golint" and commit the result.
```

thaJeztah commented:
ping @bbayani can you look at the failing CI?

cpuguy83 commented:
If we really need to support prefix matching, we should use go-patricia like we do for containers.
I don't think walking the entire volume list is desirable here.

bbayani commented:
@cpuguy83 : you commented on issue that you would not prefer go-patricia stuff as you are planning on removing it. Thus implemented the walking for entire volume list. Let me know if you have changed mind. I should implement similar prefix matching to containers.

cpuguy83 commented:
We can use it while it's there, but I'm also hoping to get a PR in soon that does indeed remove it, so it may be for naught.

jhowardmsft commented:
How can I help move this forward?

bbayani commented:
I am not clear. Should I use go-patria approach or not?

bbayani committed:
partial volume ID matching: return error if not found
Adressing golint failures
Addressing unit test failures
Addressing gofmt failure

Signed-off-by: Bhumika Bayani <bhumikabayani@gmail.com>

bbayani commented:
Also, i am not able to understand the failing unit test in docker_api_containers_test.go..
FAIL: docker_api_containers_test.go:1776  DockerSuite.TestContainersAPICreateMountsCreate

Can someone guide?

thaJeztah commented:
@bbayani the test checks if an anonymous volume is automatically removed when the container is removed. Possibly due to the prefix matching, a different volume with the same prefix (e.g. `test11`) is returned (so the test thinks the volume is not removed). See https://github.com/docker/docker/blob/master/integration-cli/docker_api_containers_test.go#L1898-L1900

bbayani commented:
Thanks @thaJeztah! The unit tests are passing now.

bbayani commented:
@thaJeztah All tests are passing now. Can you please let me know if there any more changes needed?

thaJeztah commented:
Actually, looking at this change, I wonder if this is actually correct.
For other object types (containers, images, networks), we only do partial
matching on ID's, not on _names_. Volumes in docker don't have an ID, only
a name (for anonymous volumes, a name is generated; it may "look" like an ID,
but it's actually just a random name).

For example;

```
docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
04f5e223c6ff        busybox             "sh"                6 seconds ago       Up 5 seconds                            serene_lamport
01367cafcb32        busybox             "sh"                12 seconds ago      Up 11 seconds                           gloomy_murdock
cf079da3c9e3        busybox             "sh"                7 minutes ago       Up 7 minutes                            foobar
995a06a8dc68        busybox             "sh"                7 minutes ago       Up 7 minutes                            con2
e2748f8067d1        busybox             "sh"                7 minutes ago       Up 7 minutes                            con1
```

Trying to match partial names always returns an error;

```
root@921abf8761fe:/go/src/github.com/docker/docker# docker container inspect con
[]
Error: No such container: con

root@921abf8761fe:/go/src/github.com/docker/docker# docker container inspect foo
[]
Error: No such container: con
```

Only a partial _ID_ works; 

```
docker container inspect 99
[
    {
        "Id": "995a06a8dc68b323e227e78edb3afa3ef6e2bea2bd10807a867dbd2aa26f4679",
        "Created": "2016-11-14T13:22:07.952561089Z",
        "Path": "sh",
...
```

But produces an error if multiple results were found;

```
root@921abf8761fe:/go/src/github.com/docker/docker# docker container inspect 0
[]
Error response from daemon: Multiple IDs found with provided prefix: 04f5e223c6ff201e8a2f794a6c213f088aba6379ad93c43837880a4720973145
```
(Note that the error is incorrect, as it doesn't include the partial ID provided, which is something we should address)

Same for images:

```
docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
wordpress           latest              51f253b66d9f        4 days ago          420 MB
httpd               latest              50f10ef90911        5 days ago          193 MB
busybox             latest              e02e811dd08f        5 weeks ago         1.09 MB

root@921abf8761fe:/go/src/github.com/docker/docker# docker image inspect busy
[]
Error: No such image: busy

docker image inspect 5
[]
Error: No such image: 5
```

(this error message is inconsistent, so may have to be updated)


And for networks:

```
docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
35f057677352        bridge              bridge              local
d643531bb6ae        host                host                local
dbbf8b17d60b        net1                bridge              local
b21cd179c0b3        net2                bridge              local
e8f1b9e0ad31        none                null                local
root@921abf8761fe:/go/src/github.com/docker/docker# docker network inspect net
[]
Error: No such network: net
root@921abf8761fe:/go/src/github.com/docker/docker# docker network inspect foo
[]
Error: No such network: foo

docker network inspect d
[]
Error response from daemon: invalid id: d
```

(this error message is inconsistent, so may have to be updated)

thaJeztah commented:
With the changes in this PR, partial lookup for volume _names_ is implemented;

```
docker volume ls
DRIVER              VOLUME NAME
local               db1
local               db2
```

However, if multiple matches are possible, the first one is used;

```
docker volume inspect db
[
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/db1/_data",
        "Name": "db1",
        "Options": {},
        "Scope": "local"
    }
]
```


I think that;

- we should decide if we want partial matches to work for volumes (given that
volumes don't have an ID, that would be consistent with other objects)
- if we _do_ device we want it to work, we should produce an error if multiple
results are found
- separate to this PR, we should review all the error messages, and make them
consistent (not sure which error message of we prefer)

ping @jhowardmsft @cpuguy83 WDYT?

thaJeztah commented:
I'm moving this back to design review per the above (sorry @bbayani, not your fault, just now realized we would not be consistent, so this requires additional discussion)

bbayani commented:
@thaJeztah no issues, once you guys decide on design, I can make required changes in this PR itself? Thanks!

bbayani commented:
ping @cpuguy83 @jhowardmsft @thaJeztah

cpuguy83 commented:
@bbayani Thanks for submitting this. In a maintainers meeting we decided that since we don't currently do partial matching on names of other objects it's best not to add them here, for now.

This is not to say this is forever a no, just not at this time.
Thanks again and sorry for the delay in responding to this!

jhowardmsft commented:
@thaJeztah Can this be re-visited to match by partial-`ID`?

jhowardmsft commented:
@thaJeztah Not quite true....

```
PS E:\go\src\github.com\docker\docker> docker volume create
547502f15897612c373486df3e8b3b49959ed4b6f56c504dc4ce078c4ae76f61
PS E:\go\src\github.com\docker\docker> docker volume ls
DRIVER              VOLUME NAME
local               547502f15897612c373486df3e8b3b49959ed4b6f56c504dc4ce078c4ae76f61
PS E:\go\src\github.com\docker\docker>
```

jhowardmsft commented:
Well half and half - the name is an ID. Which is horrible when it comes to attempting to delete it.

thaJeztah commented:
Yes, it's a random name, and it "looks" like an ID. I'd personally be +1 if there was a way to give volumes an ID to be consistent with all other objects we have
